### PR TITLE
[TVMScript] Text underlining in DocPrinter based on Doc's source_paths

### PR DESCRIPTION
**Pull request by gbonik**
This adds an ability to print a "diagnostic marker" based on a given ObjectPath. For example, say we are printing a fragment of TIR like
```
for i in T.serial(10):
    a[i] = 5
```
and we would like bring the user's attention to the bound of the loop:
```
for i in T.serial(10):
                  ^^
    a[i] = 5
```
In this case we would give the doc printer an object path that represents this loop bound, i.e. something like `path_to_underline=ObjectPath.root().attr("extent")`

Tracking issue: https://github.com/apache/tvm/issues/11912
cc @yelite @junrushao1994

### CODE REVIEW

File: src/script/printer/base_doc_printer.cc [hunk 1]
```diff
@@ -23,19 +23,256 @@ namespace tvm {
...
+        printed_underline = true;
+      }
+
+      size_t underline_end_for_line = std::min(line_end, current_underline.second);
+      size_t num_spaces = GetTextWidth(text, {line_pos, current_underline.first});
+      if (num_spaces > 0 && printed_extra_caret) {
+        num_spaces -= 1;
+        printed_extra_caret = false;
+      }
+      *out += std::string(num_spaces, ' ');
```
Comment (yelite) [inline src/script/printer/base_doc_printer.cc:115]: Should it skip printing carets under the indentation spaces for multi-line underline?

**Review by yelite — APPROVED** [review]

**Review by junrushao1994 — APPROVED** [review]
LGTM

**Comment by junrushao1994** [comment]
Let's fix the lint and get it in :-)

File: src/script/printer/base_doc_printer.cc (hunk 1 above)
Comment (gbonik) [reply src/script/printer/base_doc_printer.cc:115]: I'm not sure, I could see arguments either way.
```
for i in T.serial(10):
^^^^^^^^^^^^^^^^^^^^^^
    a[i] = 5
^^^^^^^^^^^^
```

```    
for i in T.serial(10):
^^^^^^^^^^^^^^^^^^^^^^
    a[i] = 5
    ^^^^^^^^
```
I'd say that the first option can provide a better sense of "continuity". make it clear that this is one chunk of text being highlighted, as opposed to two different chunks. Also it seems to simplify the implementation :)

Comment (yelite) [reply src/script/printer/base_doc_printer.cc:115]: This makes sense. What I was thinking about is the deeply nested code, like

```
                  for i in T.serial(10):
                  ^^^^^^^^^^^^^^^^^^^^^^
                     a[i] = 5
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                     b[i] = 6
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

But it does add more complexity to the implementation. We can keep it as is for now and make improvement in the future if needed.

File: src/script/printer/base_doc_printer.cc [hunk 2]
```diff
@@ -23,19 +23,256 @@ namespace tvm {
...
+
+ByteSpan PopNextUnderline(UnderlineIter* next_underline, UnderlineIter end_underline) {
+  if (*next_underline == end_underline) {
+    return {std::numeric_limits<size_t>::max(), std::numeric_limits<size_t>::max()};
+  } else {
+    return *(*next_underline)++;
+  }
+}
+
+void PrintChunk(const std::pair<size_t, size_t>& lines,
```
[the remark this answers is not in the document]
Comment (gbonik) [reply src/script/printer/base_doc_printer.cc:78]: Good call, fixed

File: src/script/printer/base_doc_printer.cc (hunk 1 above)
Comment (gbonik) [reply src/script/printer/base_doc_printer.cc:115]: Good point about the nested code. But let's give the simpler implementation a shot and then add more complexity if people hate this behavior.

**Comment by junrushao1994** [comment]
Thanks for the PR! It's merged :-)

**Pull request merged**